File: discord-bot/pre-tool-gate.js
/**
 * pre-tool-gate.js — Discord 봇 PreExecute 게이트
 *
 * 정책 SSOT: whosbuying/.agent/harness/policies/sensitive-paths.yaml
 * Python(pre_tool_gate.py) 과 동일 yaml 을 로드하여 패턴 fork 를 방지한다.
 *
 * 실패 코드:
 *   HG001: sensitive path denied
 *   HG002: dangerous command denied
 *
 * yaml 로드 실패 시 fail-open(degraded) — 인라인 fallback 패턴 사용.
 * 운영 차단보다 운영 가능성 우선 (CLAUDE.md 6조).
 */

const fs = require('fs');
const path = require('path');

// ── SSOT 로더 (의존성 없는 minimal yaml list 파서) ────────────
const POLICY_PATH = path.resolve(
  __dirname,
  '../../whosbuying/.agent/harness/policies/sensitive-paths.yaml'
);

function _extractYamlList(text, section) {
  // 라인 단위 파싱: section 헤더(`<section>:`)부터 다음 top-level key(또는 EOF)까지의
  // '- ...' 라인을 수집. 들여쓰기 있는 '- '만 항목으로 인정.
  const lines = text.split('\n');
  const out = [];
  let inSection = false;
  for (const raw of lines) {
    if (raw.match(new RegExp(`^${section}:\\s*$`))) {
      inSection = true;
      continue;
    }
    if (!inSection) continue;
    // 다음 top-level key (들여쓰기 없는 'key:') 만나면 종료
    if (/^[a-zA-Z_][a-zA-Z0-9_]*:\s*/.test(raw)) break;
    const t = raw.trim();
    if (!t || t.startsWith('#')) continue;
    if (!t.startsWith('- ') && t !== '-') continue;
    let val = t.slice(1).trim();
    // PR-GATE-EXCEPTIONS-JS — inline comment 처리 (quote 안의 # 는 보존)
    const m = val.match(/^(['"])(.*?)\1\s*(#.*)?$/);
    if (m) {
      val = m[2];
    } else {
      // 무 quote: # 이전까지만
      const hashIdx = val.indexOf('#');
      if (hashIdx !== -1) val = val.slice(0, hashIdx).trim();
    }
    if (val) out.push(val);
  }
  return out.length ? out : null;
}

/**
 * PR-GATE-EXCEPTIONS-JS — nested yaml list 파서
 * 'parent:' 안의 'child:' 의 - 항목 수집.
 * 예: exceptions.sensitive_paths
 */
function _extractNestedYamlList(text, parent, child) {
  const lines = text.split('\n');
  const subLines = [];
  let inParent = false;
  for (const raw of lines) {
    if (raw.match(new RegExp(`^${parent}:\\s*$`))) {
      inParent = true;
      continue;
    }
    if (!inParent) continue;
    // 다음 top-level key 만나면 종료
    if (/^[a-zA-Z_][a-zA-Z0-9_]*:\s*$/.test(raw)) break;
    subLines.push(raw);
  }
  if (!subLines.length) return [];

  const out = [];
  let inChild = false;
  for (const raw of subLines) {
    // child key (들여쓰기 있는 'child:') 매칭
    if (raw.match(new RegExp(`^\\s+${child}:\\s*$`))) {
      inChild = true;
      continue;
    }
    if (!inChild) continue;
    // 같은/다른 nested key 만나면 종료
    if (
      /^\s+[a-zA-Z_][a-zA-Z0-9_]*:\s*$/.test(raw) &&
      !raw.match(new RegExp(`^\\s+${child}:`))
    ) {
      break;
    }
    const t = raw.trim();
    if (!t || t.startsWith('#')) continue;
    if (!t.startsWith('- ') && t !== '-') continue;
    let val = t.slice(1).trim();
    // PR-GATE-EXCEPTIONS-JS — inline comment 처리 (quote 안의 # 는 보존)
    const m = val.match(/^(['"])(.*?)\1\s*(#.*)?$/);
    if (m) {
      val = m[2];
    } else {
      // 무 quote: # 이전까지만
      const hashIdx = val.indexOf('#');
      if (hashIdx !== -1) val = val.slice(0, hashIdx).trim();
    }
    if (val) out.push(val);
  }
  return out;
}

function _loadPolicy() {
  const fallback = {
    sensitive: [
      /\.env($|\.)/i,
      /secrets[/\\]/i,
      /credentials?[/\\]/i,
      /\.pem$/i,
      /\.key$/i,
      /id_rsa/i,
      /id_ed25519/i,
    ],
    dangerous: [
      /\brm\s+-rf\b/i,
      /\bgit\s+reset\s+--hard\b/i,
      /\bgit\s+push\s+.*--force\b/i,
      /\bsudo\s+rm\b/i,
    ],
    exceptions: [], // PR-GATE-EXCEPTIONS-JS — fallback 시 빈 리스트 (기존 동작)
  };
  try {
    const text = fs.readFileSync(POLICY_PATH, 'utf8');
    const sensitive = _extractYamlList(text, 'sensitive_paths');
    const dangerous = _extractYamlList(text, 'dangerous_commands');
    // PR-GATE-EXCEPTIONS-JS — exceptions.sensitive_paths nested 추출
    const exceptions = _extractNestedYamlList(text, 'exceptions', 'sensitive_paths');
    if (!sensitive || !dangerous) throw new Error('empty policy lists');
    return {
      sensitive: sensitive.map((p) => new RegExp(p, 'i')),
      dangerous: dangerous.map((p) => new RegExp(p, 'i')),
      exceptions: exceptions.map((p) => new RegExp(p, 'i')),
    };
  } catch (exc) {
    process.stderr.write(
      `[pre-tool-gate] policy load failed (${exc.message}); using fallback patterns\n`
    );
    return fallback;
  }
}

const _policy = _loadPolicy();
const SENSITIVE_PATTERNS = _policy.sensitive;
const DANGEROUS_PATTERNS = _policy.dangerous;
const EXCEPTION_PATTERNS = _policy.exceptions;

/**
 * PR-GATE-EXCEPTIONS-JS — false-positive 차단
 * @param {string} path
 * @returns {boolean}
 */
function _isException(path) {
  if (!path) return false;
  for (const pat of EXCEPTION_PATTERNS) {
    if (pat.test(path)) return true;
  }
  return false;
}

/**
 * 경로가 민감한 패턴과 일치하는지 확인
 * @param {string} path
 * @returns {boolean}
 */
function checkSensitivePath(path) {
  if (!path) return false;
  // PR-GATE-EXCEPTIONS-JS — exceptions 우선 검증
  if (_isException(path)) return false;
  for (const pat of SENSITIVE_PATTERNS) {
    if (pat.test(path)) return true;
  }
  return false;
}

/**
 * 명령이 파괴 패턴과 일치하는지 확인
 * @param {string} cmd
 * @returns {boolean}
 */
function checkDangerousCommand(cmd) {
  if (!cmd) return false;
  for (const pat of DANGEROUS_PATTERNS) {
    if (pat.test(cmd)) return true;
  }
  return false;
}

/**
 * 텍스트에서 경로로 보이는 부분 추출
 * @param {string} text
 * @returns {string[]}
 */
function extractPaths(text) {
  const paths = [];
  if (!text) return paths;

  // 파일 경로 패턴 추출 (예: /path/to/file, C:\path\to\file)
  const pathPatterns = [
    /[a-zA-Z]:[\w./\\-]+/g,           // Windows: C:\...
    /\/[a-zA-Z0-9._/\-]*/g,            // Unix: /path/to/...
    /\w+[\w./\-]*\.\w+/g,              // files: name.ext
  ];

  for (const pat of pathPatterns) {
    const matches = text.match(pat) || [];
    paths.push(...matches);
  }

  return paths;
}

/**
 * 사용자 메시지 검증
 * @param {string} userMessage
 * @returns {Object} { decision: 'allow'|'deny', reason: string }
 */
function validateUserMessage(userMessage) {
  if (!userMessage || typeof userMessage !== 'string') {
    return { decision: 'allow', reason: 'empty message' };
  }

  // ── 민감 경로 검사 ──
  const paths = extractPaths(userMessage);
  for (const p of paths) {
    if (checkSensitivePath(p)) {
      return {
        decision: 'deny',
        reason: `HG001: sensitive path denied — ${p}`,
      };
    }
  }

  // ── 파괴 명령 검사 ──
  if (checkDangerousCommand(userMessage)) {
    return {
      decision: 'deny',
      reason: `HG002: dangerous command denied — ${userMessage.substring(0, 120)}`,
    };
  }

  // ── 통과 ──
  return { decision: 'allow', reason: 'passed all gates' };
}

module.exports = {
  validateUserMessage,
  checkSensitivePath,
  checkDangerousCommand,
  // 패턴 노출(검증/diff 용도)
  SENSITIVE_PATTERNS,
  DANGEROUS_PATTERNS,
};

File: .claude/hooks/pre_tool_gate.py
#!/usr/bin/env python3
"""
pre_tool_gate.py — Claude Code PreToolUse hook.
민감 경로 접근과 파괴 명령을 차단한다.

실패 코드:
  HG001: sensitive path denied
  HG002: dangerous command denied

stdin: Claude Code hook JSON (tool_name, tool_input 등)
stdout: JSON {"decision": "allow"|"deny", "reason": "..."}
"""

import json
import sys
import re
import os
from datetime import datetime, timezone
from pathlib import Path
from typing import List

# ── SSOT: sensitive-paths.yaml 로드 ─────────────────────────
# 정책 SSOT 는 .agent/harness/policies/sensitive-paths.yaml 이다.
# yaml 로드 실패 시 fail-open(degraded allow) — 인라인 fallback 패턴 사용.
# 운영 차단보다 운영 가능성 우선 (CLAUDE.md 6조).

_FALLBACK_SENSITIVE = [
    r"\.env($|\.)",
    r"secrets[/\\]",
    r"credentials?[/\\]",
    r"\.pem$",
    r"\.key$",
    r"id_rsa",
    r"id_ed25519",
]
_FALLBACK_DANGEROUS = [
    r"\brm\s+-rf\b",
    r"\bgit\s+reset\s+--hard\b",
    r"\bgit\s+push\s+.*--force\b",
    r"\bsudo\s+rm\b",
]


def _load_policy():
    """sensitive-paths.yaml 로드 → (sensitive[], dangerous[], exceptions[]) 반환.
    실패 시 fallback 패턴 반환 + stderr 경고.
    PR-GATE-EXCEPTIONS — exceptions 섹션 추가 (false-positive 차단)."""
    repo_root = Path(__file__).resolve().parents[2]
    policy_path = repo_root / ".agent" / "harness" / "policies" / "sensitive-paths.yaml"
    try:
        import yaml  # type: ignore
        with policy_path.open("r", encoding="utf-8") as f:
            data = yaml.safe_load(f) or {}
        sensitive = list(data.get("sensitive_paths") or [])
        dangerous = list(data.get("dangerous_commands") or [])
        # exceptions: 미존재 시 빈 리스트 (기존 동작 유지 — backward compat)
        exceptions_section = data.get("exceptions") or {}
        path_exceptions = list(exceptions_section.get("sensitive_paths") or [])
        if not sensitive or not dangerous:
            raise ValueError("empty policy lists")
        return sensitive, dangerous, path_exceptions
    except Exception as exc:
        sys.stderr.write(
            f"[pre_tool_gate] policy load failed ({exc}); using fallback patterns\n"
        )
        return _FALLBACK_SENSITIVE, _FALLBACK_DANGEROUS, []


SENSITIVE_PATTERNS, DANGEROUS_COMMANDS, PATH_EXCEPTIONS = _load_policy()
SENSITIVE_RE = [re.compile(p, re.IGNORECASE) for p in SENSITIVE_PATTERNS]
DANGEROUS_RE = [re.compile(p, re.IGNORECASE) for p in DANGEROUS_COMMANDS]
PATH_EXCEPTION_RE = [re.compile(p, re.IGNORECASE) for p in PATH_EXCEPTIONS]


def _is_exception(path: str) -> bool:
    """PR-GATE-EXCEPTIONS — 정당한 false-positive 예외 매치."""
    if not path:
        return False
    for pat in PATH_EXCEPTION_RE:
        if pat.search(path):
            return True
    return False


def check_sensitive_path(path: str) -> bool:
    if not path:
        return False
    # exceptions 우선 — false-positive 차단
    if _is_exception(path):
        return False
    for pat in SENSITIVE_RE:
        if pat.search(path):
            return True
    return False


def check_dangerous_command(cmd: str) -> bool:
    if not cmd:
        return False
    for pat in DANGEROUS_RE:
        if pat.search(cmd):
            return True
    return False


def extract_paths(tool_input: dict) -> List[str]:
    """tool_input에서 경로로 보이는 값을 추출한다."""
    paths = []
    for key in ("file_path", "path", "file", "directory", "glob", "pattern"):
        val = tool_input.get(key)
        if isinstance(val, str) and val:
            paths.append(val)
    # command 안에 포함된 경로도 검사
    cmd = tool_input.get("command", "")
    if isinstance(cmd, str):
        paths.append(cmd)
    return paths


def main():
    try:
        raw = sys.stdin.read()
        if not raw.strip():
            # 입력 없으면 allow
            print(json.dumps({"decision": "allow", "reason": "no input"}))
            return

        event = json.loads(raw)
    except (json.JSONDecodeError, Exception):
        # 파싱 실패 시 allow (logger 성격과 달리 gate는 보수적 deny도 가능하지만,
        # 파싱 실패로 모든 작업을 막으면 운영 불가이므로 allow + 경고)
        print(json.dumps({
            "decision": "allow",
            "reason": "hook input parse failed — degraded allow"
        }))
        return

    tool_name = event.get("tool_name", "")
    tool_input = event.get("tool_input", {})
    if isinstance(tool_input, str):
        try:
            tool_input = json.loads(tool_input)
        except json.JSONDecodeError:
            tool_input = {"command": tool_input}

    # ── 민감 경로 검사 ──
    for p in extract_paths(tool_input):
        if check_sensitive_path(p):
            result = {
                "decision": "deny",
                "reason": f"HG001: sensitive path denied — {p}"
            }
            print(json.dumps(result))
            return

    # ── 파괴 명령 검사 ──
    cmd = tool_input.get("command", "")
    if isinstance(cmd, str) and check_dangerous_command(cmd):
        result = {
            "decision": "deny",
            "reason": f"HG002: dangerous command denied — {cmd[:120]}"
        }
        print(json.dumps(result))
        return

    # ── 통과 ──
    print(json.dumps({"decision": "allow", "reason": "passed all gates"}))


if __name__ == "__main__":
    main()
